Add cohorts to BulkEnroll endpoint.

pomegranited opened the pull request:
This PR adds the ability to assign a cohort to the users from the BulkEnroll endpoint.

**JIRA tickets**: SE-15, [OSPR-2642](https://openedx.atlassian.net/browse/OSPR-2642)

**Discussions**: Replacement PR for https://github.com/edx/edx-platform/pull/18282

**Dependencies**: None

**Screenshots**: None -- no GUI changes made.

**Sandbox URL**:  

Running 5896d98 

* LMS: https://pr19032.sandbox.opencraft.hosting/
* Studio: https://studio-pr19032.sandbox.opencraft.hosting/

For testing purposes:
* A course has already been created on the sandbox: https://pr19032.sandbox.opencraft.hosting/courses/course-v1:OpenCraft+cohorts+2018_1
* Two cohorts "Apples" and "Oranges" have been added to the test course.
* An OAuth2 token `fake-token` has been generated 

**Merge deadline**: None

**Testing instructions**:

1. Create a new course and add cohorts to it. cf [cohorts docs](http://edx.readthedocs.io/projects/edx-partner-course-staff/en/latest/course_features/cohorts/cohorts_overview.html#cohorts-overview)
1. Hit the bulk enroll endpoint with the correct data and the learner should be enrolled in the course, as well as added to the required cohort.
  E.g., the following command:
   ```bash
   curl -X POST --header "Authorization: Bearer fake-token" -H "Content-Type: application/json" \
      -d '{"action": "enroll","auto_enroll": true,"email_students": true,"courses": "course-v1:OpenCraft+cohorts+2018_1", "identifiers": "verified", "cohorts": "Apples"}' \
      https://pr19032.sandbox.opencraft.hosting/api/bulk_enroll/v1/bulk_enroll
   ```
   Running this command [enrolls the `verified` user in the `Apples` cohort](https://pr19032.sandbox.opencraft.hosting/courses/course-v1:OpenCraft+cohorts+2018_1/instructor#view-cohort_management), and returns:
   ```javascript
   {
      "action":"enroll",
      "courses":{
         "course-v1:OpenCraft+cohorts+2018_1":{
            "action":"enroll",
            "results":[
               {
                  "identifier":"verified",
                  "after":{
                     "cohort":"Apples",
                     "enrollment":true,
                     "allowed":false,
                     "user":true,
                     "auto_enroll":false
                  },
                  "before":{
                     "cohort":null,
                     "enrollment":false,
                     "allowed":false,
                     "user":true,
                     "auto_enroll":false
                  }
               }
            ],
            "auto_enroll":true
         }
      },
      "email_students":true,
      "auto_enroll":true
   }
   ```
1. An invalid call will return an error.  
  E.g. the number of cohorts and number of courses must match:
   ```bash
   curl -X POST --header "Authorization: Bearer fake-token" -H "Content-Type: application/json" \
      -d '{"action": "enroll","auto_enroll": true,"email_students": true,"courses": "course-v1:OpenCraft+cohorts+2018_1","identifiers": "verified", "cohorts": "Apples,Oranges"}' \
      https://pr19032.sandbox.opencraft.hosting/api/bulk_enroll/v1/bulk_enroll
   ```
   returns:
   ```javascript
   {
      "non_field_errors": ["If provided, the cohorts and courses should have equal number of items."]
   }
   ```

**Reviewers**
- [ ] @symbolist 
- [ ] edX reviewer[s] TBD

**Settings**
```yaml
EDXAPP_FEATURES:
  ENABLE_BULK_ENROLLMENT_VIEW: true
  ENABLE_COMBINED_LOGIN_REGISTRATION: true
```

edx-webhook commented:
Thanks for the pull request, @pomegranited! I've created [OSPR-2642](https://openedx.atlassian.net/browse/OSPR-2642) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

pomegranited commented:
jenkins run python

natabene commented:
Hi @pomegranited This is great, thank you for your contribution.  @edx/arch-review can you have a look at this?

mduboseedx commented:
@pomegranited We also had some problems with our tests. Could you rebase this and try the tests again?

pomegranited commented:
Brilliant, that did the trick, thanks @mduboseedx !

@natabene You can thank @afzaledx and @symbolist for this code -- it's their work, all I did was resubmit the original PR 😄

afzaledx commented:
Hey @natabene, I didn't know you looked at PRs as well. Great to find you here.

nasthagiri commented:
@davestgermain Since you recently updated the Enrollment API to include the cohort as a parameter, can you take a look at this PR to see if it is directionally aligned with your work?

pomegranited commented:
@davestgermain CC @nasthagiri @mduboseedx Do you know when you'll have a chance to look at this PR?

davestgermain commented:
@pomegranited I think it looks good, but could you address my two comments above?

pomegranited commented:
Sorry @davestgermain -- I can't find your comments, do you mind to repost them?  Maybe they got lost in a squashed commit or something?

pomegranited commented:
Thank you @davestgermain !  I don't have time to address your feedback this sprint, but will do this next sprint (Nov 6).

pomegranited commented:
Thanks for your patience and recommendations here @davestgermain!  I've rebased f9e11a9 on the latest `master` and addressed your comments with 5896d98c.

Am rebuilding the sandbox now.  Let me know if you see anything else that needs fixing before this can be merged?

davestgermain commented on lms/djangoapps/bulk_enroll/views.py:
```diff
@@ -77,7 +76,6 @@ class BulkEnrollView(APIView):
     throttle_classes = (EnrollmentUserThrottle,)
 
-    @method_decorator(transaction.non_atomic_requests)
     def dispatch(self, request, *args, **kwargs):
```
Since you removed the decorator, this method isn't needed.

davestgermain commented:
@natabene Yes, but it would be nice if the commits were squashed first.

👍

pomegranited commented on lms/djangoapps/bulk_enroll/views.py:
```diff
@@ -51,12 +67,19 @@ class BulkEnrollView(APIView):
     permission_classes = IsStaff,
     throttle_classes = EnrollmentUserThrottle,
 
+    @method_decorator(transaction.non_atomic_requests)
```
Yep, quality caught that too, thanks!

pomegranited commented:
Thank you for your review @davestgermain, and to @natabene for keeping tabs on the status !  I've squashed the commits, so this should be ok to merge when tests pass.

natabene commented:
@pomegranited  Thank you! @davestgermain  Woohoo, it is ready now.

edx-webhook commented:
@pomegranited 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/19032&entry.1484655318=https://github.com/pomegranited&entry.752974735=2018-10-02+04:31&entry.1917517419=2018-11-08+16:36&entry.2133058324=Yes) so we can improve your experience in the future.
